_drafts/2020-09-17-fixing-an-n1-performance-problem-in-xaf-xpo-with-totally-undocumented-apis.md

```diff
@@ -0,0 +1,948 @@
+19.09.20 12:19:10.993 Result: rowcount = 55, total = 3108, SubQuery(Sum,N1.{Hours,Int32},Select N1.{Hours,Int32}
+  from "SlowOfferItemWithPersistentAlias" N1
+ where N0.{Oid,Guid} = N1.{SlowOfferWithPersistentAlias,Guid} And N1.{GCRecord,Int32} Is Null) = 220, N0.{Name,String} = 908, N0.{Oid,Guid} = 880, N0.{Oid,Guid} = 880, SubQuery(Sum,N2.{Hours,Int32},Select N2.{Hours,Int32}
+  from "SlowOfferItemWithPersistentAlias" N2
+ where N0.{Oid,Guid} = N2.{SlowOfferWithPersistentAlias,Guid} And N2.{GCRecord,Int32} Is Null) = 220
+19.09.20 12:19:10.995 Executing sql 'select top 1 count(*) from "dbo"."SlowOfferWithPersistentAlias" N0 where N0."GCRecord" is null'
+19.09.20 12:19:10.995 Result: rowcount = 1, total = 4, SubQuery(Count,,) = 4
+19.09.20 12:19:10.996 Executing sql 'select top 1 count(*) from "dbo"."SlowOfferItemWithPersistentAlias" N0 where N0."GCRecord" is null'
+19.09.20 12:19:11.010 Result: rowcount = 1, total = 4, SubQuery(Count,,) = 4
+```
+
+This results in the following performance
+
+1. Sort by `Name`: 0.079 seconds
+2. Sort by `HourSum`: 0.236 seconds
+
+> **Important**: When using `PersistentAlias` for aggregate calculations always consider using a `DataAccessMode` that can perform database side calculation.  
+There is no silverbullet and you need to measure careful in different configurations what fit's best.
+
+All comes with a cost. All data access modes have [some limitations](https://docs.devexpress.com/eXpressAppFramework/113683/concepts/ui-construction/views/list-view-data-access-modes#non-persistent-object-support-limitations). If you need grouping and sorting capabilities by some fields (or even just display `NonPersistent` members) you are *out of luck*. It is also *may* be limited to the complexity you can fit into a single criteria.  
+
+## Techniques that overcome the N+1 problems
+
+Now we learned about the N+1 problem and some useful ways to overcome some of the performance problems that XAF provides we will cover 2 supported and 1 unsupported techniques how to avoid the N+1 problem completely.  
```

Comment:
```suggestion
Now we learned about the N+1 problem and some useful ways to overcome some of the performance problems that XAF causes. We will cover 2 supported and 1 unsupported techniques to avoid the N+1 problem completely.  
```

Reply:
*can* cause.
XAF cant know what we want. Poor XAF ❤️